site/docs/relations/logical_relations.md

```diff
@@ -219,6 +219,40 @@ The join operation will combine two separate inputs into a single output, based
     ```
 
 
+## AsOfJoin Operation
+
+The as-of-join operation is a time-series operation that will combine a left input and multiple right inputs into a single output, based on a join expression, an on-field and a tolerance value. All inputs must have the on-field in ascending-order. The operation is similar to a join-operation where the join expression is used for exact matching whereas the on-field is used for inexact matching up to the tolerance value.
+
+| Signature            | Value                                                        |
+| -------------------- | ------------------------------------------------------------ |
+| Inputs               | 2 or more                                                    |
+| Outputs              | 1                                                            |
+| Property Maintenance | Distribution is maintained. Orderedness is by the on-field only post operation. Physical relations may provide better property maintenance. |
+| Direct Output Order  | The emit order of the left input followed by the emit order of each right input. |
+
+### Join Properties
+
+| Property        | Description                                                  | Required                           |
+| --------------- | ------------------------------------------------------------ | ---------------------------------- |
+| Left Input      | A relational input.                                          | Required                           |
+| Right Inputs    | Each a relational input.                                     | Required, at least one             |
+| Join Expression | A boolean condition that describes whether each record from the left set "match" the record from the right set. Field references correspond to the direct output order of the data. | Required. Can be the literal True. |
+| Join Type       | Same as the [Join](logical_relations.md#join-operator) operator. | Required                           |
+| Tolerance       | The maximum on-field value difference in an inexact match.   | Required                           |
+| On              | The on-field.                                                | Required                           |
+
+### Join Types
+
+Same as in the join operation.
```

Comment:
I don't think these are all properties of the AsOf join (and an AsOf join, if I am reading the protobuf correctly, doesn't have a join type).  So I think all you need is something like...

| Property | ...
| Join | ...
| Tolerance | ...
| On | ...

Reply:
I define here a generalized as-of-join operation, which is a generalization of both the join operation and the current `AsofJoinNode` in Arrow (more details below). My understanding here is that it is OK (and desirable, I think) to define a more general logical operation, even if it is currently only partially implemented in physical operations, provided the general operation is well-designed.

At a high level, the table result of the generalized as-of-join operation, with given join-expression and join-type, is a (not necessarily struct) sub-table of the join-operation with the same join-expression and join-type. The site-doc for as-of-join says that a row in the join-result will also appear in the as-of-join-result if the inexact matching of the on-key evaluates to true for this row. This is why any join-type also works in the generalized as-of-join operation. Note that the `AsOfJoinNode` in Arrow is equivalent to the generalized as-of-join operation with a left-join as the join-type.

Reply:
It's fine for it to be more general in principle, and for implementations to support the various join types with as of join, but let's keep the implementation decoupled from `JoinRel`.

Reply:
You mean just duplicate the `JoinRel` fields into `AsOfJoinRel`?